ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/altair/statetransition/epoch/RewardsAndPenaltiesCalculatorAltair.java

```diff
@@ -0,0 +1,203 @@
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.spec.config.SpecConfigAltair;
+import tech.pegasys.teku.spec.constants.ParticipationFlags;
+import tech.pegasys.teku.spec.datastructures.state.beaconstate.versions.altair.BeaconStateAltair;
+import tech.pegasys.teku.spec.logic.common.statetransition.epoch.RewardAndPenaltyDeltas;
+import tech.pegasys.teku.spec.logic.common.statetransition.epoch.RewardAndPenaltyDeltas.RewardAndPenalty;
+import tech.pegasys.teku.spec.logic.common.statetransition.epoch.RewardsAndPenaltiesCalculator;
+import tech.pegasys.teku.spec.logic.common.statetransition.epoch.status.TotalBalances;
+import tech.pegasys.teku.spec.logic.common.statetransition.epoch.status.ValidatorStatus;
+import tech.pegasys.teku.spec.logic.common.statetransition.epoch.status.ValidatorStatuses;
+import tech.pegasys.teku.spec.logic.versions.altair.helpers.BeaconStateAccessorsAltair;
+import tech.pegasys.teku.spec.logic.versions.altair.helpers.MiscHelpersAltair;
+import tech.pegasys.teku.spec.logic.versions.altair.helpers.MiscHelpersAltair.FlagIndexAndWeight;
+
+public class RewardsAndPenaltiesCalculatorAltair extends RewardsAndPenaltiesCalculator {
+
+  private final SpecConfigAltair specConfigAltair;
+  private final BeaconStateAccessorsAltair beaconStateAccessorsAltair;
+  private final MiscHelpersAltair miscHelpersAltair;
+
+  private final BeaconStateAltair stateAltair;
+  private final Map<Integer, UInt64> baseRewardCache = new HashMap<>();
+
+  public RewardsAndPenaltiesCalculatorAltair(
+      final SpecConfigAltair specConfig,
+      final BeaconStateAltair state,
+      final ValidatorStatuses validatorStatuses,
+      final MiscHelpersAltair miscHelpers,
+      final BeaconStateAccessorsAltair beaconStateAccessors) {
+    super(specConfig, miscHelpers, beaconStateAccessors, state, validatorStatuses);
+    this.stateAltair = state;
+    this.specConfigAltair = specConfig;
+    this.miscHelpersAltair = miscHelpers;
+    this.beaconStateAccessorsAltair = beaconStateAccessors;
+  }
+
+  /**
+   * Return attestation reward/penalty deltas for each validator
+   *
+   * @return
+   * @throws IllegalArgumentException
+   */
+  @Override
+  public RewardAndPenaltyDeltas getDeltas() throws IllegalArgumentException {
+    final RewardAndPenaltyDeltas deltas =
+        new RewardAndPenaltyDeltas(validatorStatuses.getValidatorCount());
+
+    for (FlagIndexAndWeight flagIndicesAndWeight : miscHelpersAltair.getFlagIndicesAndWeights()) {
+      processFlagIndexDeltas(deltas, flagIndicesAndWeight);
+    }
+    processInactivityPenaltyDeltas(deltas);
+
+    return deltas;
+  }
+
+  /**
+   * Corresponds to altair beacon chain accessor get_flag_index_deltas
+   *
+   * @see <a
+   *     href="https://github.com/ethereum/eth2.0-specs/blob/master/specs/altair/beacon-chain.md#beacon-state-accessors">Altair
+   *     beacon-chain.md</a>
```

Comment:
These links have tended to not age well. I'd be tempted to omit it.

Reply:
True, although I think the spec is much more stable than it used to be.  And now that we're live, things can't just change arbitrarily.  It was a little hard to map some of the phase0 logic to the spec, so I think it does help having these notes when we're working on updates.